apollo-router/src/plugins/authentication/jwks.rs

```diff
@@ -256,3 +273,419 @@ impl Iterator for Iter<'_> {
+        Source::Cookie { name } => {
+            for header in headers.get_all("cookie") {
+                let value = match header.to_str() {
+                    Ok(value) => value,
+                    Err(_not_a_string_error) => {
+                        return Some(Err(AuthenticationError::CannotConvertToString));
+                    }
+                };
+                for cookie in cookie::Cookie::split_parse(value) {
+                    match cookie {
+                        Err(_) => continue,
+                        Ok(cookie) => {
+                            if cookie.name() == name {
+                                if let Some(value) = cookie.value_raw() {
+                                    return Some(Ok(value));
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+
+            None
+        }
+    }
+}
+
+pub(super) fn decode_jwt(
+    jwt: &str,
+    keys: Vec<(Option<String>, Jwk)>,
+    criteria: JWTCriteria,
+) -> Result<(Option<String>, TokenData<serde_json::Value>), (AuthenticationError, StatusCode)> {
+    let mut error = None;
+    for (issuer, jwk) in keys.into_iter() {
+        let decoding_key = match DecodingKey::from_jwk(&jwk) {
+            Ok(k) => k,
+            Err(e) => {
+                error = Some((
+                    AuthenticationError::CannotCreateDecodingKey(e),
+                    StatusCode::INTERNAL_SERVER_ERROR,
+                ));
+                continue;
+            }
+        };
+
+        let key_algorithm = match jwk.common.key_algorithm {
+            Some(a) => a,
+            None => {
+                error = Some((
+                    AuthenticationError::JWKHasNoAlgorithm,
+                    StatusCode::INTERNAL_SERVER_ERROR,
+                ));
+                continue;
+            }
+        };
+
+        let algorithm = match convert_key_algorithm(key_algorithm) {
+            Some(a) => a,
+            None => {
+                error = Some((
+                    AuthenticationError::UnsupportedKeyAlgorithm(key_algorithm),
+                    StatusCode::INTERNAL_SERVER_ERROR,
+                ));
+                continue;
+            }
+        };
+
+        let mut validation = Validation::new(algorithm);
+        validation.validate_nbf = true;
+        // if set to true, it will reject tokens containing an `aud` claim if the validation does not specify an audience
+        // we don't validate audience yet, so this is deactivated
+        validation.validate_aud = false;
+
+        match decode::<serde_json::Value>(jwt, &decoding_key, &validation) {
+            Ok(v) => return Ok((issuer, v)),
+            Err(e) => {
+                tracing::trace!("JWT decoding failed with error `{e}`");
+                error = Some((
+                    AuthenticationError::CannotDecodeJWT(e),
+                    StatusCode::UNAUTHORIZED,
+                ));
```

Comment:
`error` could potentially be overridden by another one here. Is it expected ? Should we have a Vec of error instead to keep all of them ?

Reply:
I moved this code but did not write it. As such, I'm not sure what the best thing to do is.

Reply:
Oh ok, sorry. Unfortunately it hasn't be done before :p now it's like dark magic